addons/mail/static/src/js/wysiwyg.js

```diff
@@ -0,0 +1,80 @@
+/** @odoo-module **/
+
+import { renderToElement } from "@web/core/utils/render";
+import { Wysiwyg } from "@web_editor/js/wysiwyg/wysiwyg";
+import { closestBlock, setCursorEnd } from "@web_editor/js/editor/odoo-editor/src/OdooEditor";
+import { patch } from "@web/core/utils/patch";
+import { MentionList } from "@mail/core/web/mention_list";
+import { url } from "@web/core/utils/urls";
+import { usePopover } from "@web/core/popover/popover_hook";
+import { isEventHandled } from "@web/core/utils/misc";
+
+patch(Wysiwyg.prototype, {
+    setup() {
+        super.setup();
+        if (this.props.options.recordInfo?.res_model === "mail.compose.message") {
+            this.mentionList = usePopover(MentionList, {
+                position: `bottom-start`,
+                onClose: () => {
+                    this.focus();
+                },
+            });
+            this.triggerMentionList = this.triggerMentionList.bind(this);
+        }
+    },
+    async startEdition() {
+        const res = await super.startEdition(...arguments);
+        // Only enable mention list in full chatter mode web editor
+        if (this.options.recordInfo?.res_model === "mail.compose.message") {
+            this.odooEditor.document.addEventListener("keydown", this.triggerMentionList, true);
+            this.odooEditor.document.addEventListener("click", this.triggerMentionList, true);
+        }
+        return res;
+    },
+    destroy() {
+        super.destroy();
+        if (this.options.recordInfo?.res_model === "mail.compose.message" && this.odooEditor) {
+            this.odooEditor.document.removeEventListener("keydown", this.triggerMentionList, true);
+            this.odooEditor.document.removeEventListener("click", this.triggerMentionList, true);
+        }
+    },
+    async triggerMentionList(ev) {
+        await new Promise((resolve) => setTimeout(resolve, 0));
+        const selection = this.odooEditor.document.getSelection();
+        if (
+            this.isSelectionInEditable() &&
+            selection.isCollapsed &&
+            selection.rangeCount &&
+            !this.mentionList.isOpen &&
+            (ev.key === "@" || ev.key === "#")
+        ) {
+            this.stepBeforeMention = this.odooEditor._historySteps.length - 2;
+            const closest = closestBlock(this.odooEditor.document.getSelection().anchorNode);
+            this.mentionList.open(closest, {
+                type: ev.key === "@" ? "partner" : "channel",
+                onSelect: this.selectMention.bind(this),
+            });
+        } else if (!isEventHandled(ev, "MentionList.onKeydown")) {
+            this.mentionList.close();
+        }
+    },
+    selectMention: function (ev, option) {
```

Comment:
I don't see the point of doing this. The reason for using a `patch` is to achieve domain-specific functions. As what `Knowledge` did, all those functions should be in `Wysiwyg` itself. On the one hand, it is easier for others to find out how these functions work; on the other hand, no other module uses those functions. Those functions are designed for `Wysiwyg`, so even if they are domain-specific, I prefer to put them within `Wysiwyg` rather than other services unless we find those services used by others.

Reply:
At first no one will use them, but future IMPs that use the wysiwyg may see these functions and have the wrong impression that they are core functions of the wysiwyg, when they are not. I also expect this can be frustrating, because they look like they should work on all wysiwygs, when it only works on `"mail.compose.message"` because the dependencies `mentionList` is properly setup.

`Wysiwyg` is shared among all wysiwyg in Odoo. The mention list feature only works in `"mail.compose.message"`. If we are ok to keep the code in `Wysiwyg`, at the very least it should be easy to use mention list for other future uses. That would mean easily patchable function to register a new model name, and also some ways to customise the `triggerMentionList` and `selectMention` methods without risks to shadowing the current implementation for `mail.composer.message`... Or if we don't want to make that effort, either namespace the methods further or move code to some other objects like a service.

Putting things in `Wysiwyg` makes life easier for the IMP of this PR at the detriment of future IMPs on the Wysiwyg. It forces other developers to adapt this code to be reusable... or even encourages hacking around this implementation.

I don't know what's the stand for addons on the wysiwyg. @dmo-odoo are you ok with patching `Wysiwyg.prototype` and add a very simple method named `selectMention` whose implementation is very specific? I see that `knowledge` does that, which is quite surprising for me considering it adds knowledge-specific behaviour to wysiwyg too.

Reply:
Ok, thanks Alex for the explanation! I understand the meaning of using services rather than directly patching to the `Wysiwyg`. Based on that, I have an idea in mind to move these codes into `useDiscussMentionList` hook, which is only valid when the `model` is `"mail.compose.message"`. I would like to know how @dmo-odoo thinks about this. Would you prefer a more direct way as of now or a more "Discuss-related" way of using a hook and patch when needed?

Reply:
I would do the simplest way without adding new methods to Wysiwyg: patch `setup()` `startEdition()` as needed, which define specific branching for our model. Then for the rest of behaviour, encapsulate all into a custom hook for our need. No need to over-engineer the hook when it's only used for this specific need at the moment.

Reply:
I had a look at this and I would suggest keeping this code like this.
I tried to put `triggerMentionList` and `selectMention` into the service but it really relies on the `wysiwyg` and I don't think it is a good service if it cannot work correctly without a specific component, which means not able to be reused and thus, no need to make it an independent service.
If we want to make it a generic service, it will be over-engineered as you said. And there are few patches to `wysiwyg`, I don't really think the case that shadowing the current implementation of these methods will happen, especially when those methods are easy to read and specifically developed and used in `Discuss`.